plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/partition/HudiPartitionInfoLoader.java

```diff
@@ -13,110 +13,77 @@
-                .collect(Collectors.toList());
+        HoodieTimer timer = new HoodieTimer().startTimer();
 
-        // empty partitioned table
-        if (hudiPartitionInfoList.isEmpty()) {
-            return;
-        }
-
-        // non-partitioned table
-        if (hudiPartitionInfoList.size() == 1 && hudiPartitionInfoList.get(0).getHivePartitionName().isEmpty()) {
-            partitionQueue.addAll(hudiPartitionInfoList);
-            return;
-        }
+        while (isRunning || !partitionQueue.isEmpty()) {
+            String partitionName = partitionQueue.poll();
 
-        boolean shouldUseHiveMetastore = hudiPartitionInfoList.get(0) instanceof HiveHudiPartitionInfo;
-        Iterator<HudiPartitionInfo> iterator = hudiPartitionInfoList.iterator();
-        while (iterator.hasNext()) {
-            int batchSize = updateBatchSize();
-            List<HudiPartitionInfo> partitionInfoBatch = new ArrayList<>();
-            while (iterator.hasNext() && batchSize > 0) {
-                partitionInfoBatch.add(iterator.next());
-                batchSize--;
-            }
-
-            if (!partitionInfoBatch.isEmpty()) {
-                if (shouldUseHiveMetastore) {
-                    Map<String, Optional<Partition>> partitions = hudiDirectoryLister.getPartitions(partitionInfoBatch.stream()
-                            .map(HudiPartitionInfo::getHivePartitionName)
-                            .collect(Collectors.toList()));
-                    for (HudiPartitionInfo partitionInfo : partitionInfoBatch) {
-                        String hivePartitionName = partitionInfo.getHivePartitionName();
-                        if (!partitions.containsKey(hivePartitionName)) {
-                            throw new HoodieIOException("Partition does not exist: " + hivePartitionName);
-                        }
-                        partitionInfo.loadPartitionInfo(partitions.get(hivePartitionName));
-                        partitionQueue.add(partitionInfo);
-                    }
-                }
-                else {
-                    for (HudiPartitionInfo partitionInfo : partitionInfoBatch) {
-                        partitionInfo.getHivePartitionKeys();
-                        partitionQueue.add(partitionInfo);
-                    }
-                }
+            if (partitionName != null) {
+                generateSplitsFromPartition(partitionName);
             }
         }
+        log.debug("HudiPartitionInfoLoader %s finishes in %d ms", this, timer.endTimer());
     }
 
-    public Deque<HudiPartitionInfo> getPartitionQueue()
+    private void generateSplitsFromPartition(String partitionName)
     {
-        return partitionQueue;
+        Optional<HudiPartitionInfo> partitionInfo = hudiDirectoryLister.getPartitionInfo(partitionName);
+        partitionInfo.ifPresent(hudiPartitionInfo -> {
+            if (hudiPartitionInfo.getHivePartitionKeys().isEmpty() || hudiPartitionInfo.doesMatchPredicates()) {
```

Comment:
Can we run `doesMatchPredicates()` before `getHivePartitionKeys().isEmpty()`?
I found this `getHivePartitionKeys()` needs to access to Hive and it takes some network time and makes Hive have more loading. On the other hands, `doesMatchPredicates()` is running to verify the partition internally in CPU (which is faster than the network time) and there is no need to access to Hive anymore if the partition doesn't match the predicates.

Reply:
I'm trying to run again from your branch and putting the log into `ThriftHiveMetastore.getPartition` to check if we have any calls into Hive. It seems like Hive has been called for many time (as the number of partitions).
![image](https://github.com/trinodb/trino/assets/6355792/15da3b77-35a8-4a16-8108-4e6a299d3ecb)

My example is I have a table `xxx.yyy` with column "date" as partition which have more than 500 days of data. When I run `SELECT * FROM hudi.xxx.yyy WHERE date >= '2023-06-01' LIMIT 10`. It tries to call `ThriftHiveMetastore.getPartition` for every partition, but, what we expected, it should call for only around 6-7 partitions.
![image](https://github.com/trinodb/trino/assets/6355792/246be339-70ad-45f0-bc4f-f085b04edbd5)

Another question, in the case that `getHivePartitionKeys()` returns empty, this line will become true and then generate split for empty partition key(?), is this what we expected?

Reply:
Sorry, I misunderstood your point earlier. Let me take a look and fix this today.

Reply:
@ksoullpwk I've fixed the issue now. Thanks for pointing it out.

Reply:
Hi @codope, I have tested again after swapping the condition and it seems like the issue still happens. It might be something wrong in this line.
From my understanding, we want to generate the split only if the predicate is in our filter and the partition does really exists. I guess this line should be `hudiPartitionInfo.doesMatchPredicates() && !hudiPartitionInfo.getHivePartitionKeys().isEmpty()`.
So, we check first if it does match predicates, in case of yes, we will call to Hive and check if the partition key does not empty.

Reply:
@ksoullpwk After flipping the conditional expressions, metastore call should only happen for the partitions that match predicate. The second part OR condition i.e. `hudiPartitionInfo.getHivePartitionKeys().isEmpty()` is to generate splits for non-partitioned table. It was the same way before:
https://github.com/trinodb/trino/blob/5295f3d5a66020f735e993911f743a5df860269f/plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/query/HudiReadOptimizedDirectoryLister.java#L90-L92

You can also check `TestHudiSmokeTest#testReadNonPartitionedTable`

Reply:
@ksoullpwk Good catch! I have fixed the check just now. Please take a look.

Reply:
Thanks for your help.